lib/src/package_config_impl.dart

```diff
@@ -375,122 +399,127 @@ abstract class PackageTree {
 /// A package is said to be inside another package if the root path URI of
 /// the latter is a prefix of the root path URI of the former.
 ///
-/// No two packages of a package may have the same root path, so this
-/// path prefix ordering defines a tree-like partial ordering on packages
-/// of a configuration.
-///
+/// No two packages of a package may have the same root path.
 /// The package root path of a package must not be inside another package's
 /// root path.
-/// Entire other packages are allowed inside a package's root or
-/// package root path.
-///
-/// The package tree contains an ordered mapping of unrelated packages
-/// (represented by their name) to their immediately nested packages' names.
-class MutablePackageTree implements PackageTree {
-  /// A list of packages that are not nested inside each other.
-  final List<SimplePackage> packages = [];
+/// Entire other packages are allowed inside a package's root.
+class TriePackageTree implements PackageTree {
+  /// Indexed by URI scheme.
+  final Map<String, _PackageTrieNode> _map = {};
 
-  /// The tree of the immediately nested packages inside each package.
-  ///
-  /// Indexed by [Package.name].
-  /// If a package has no nested packages (which is most often the case),
-  /// there is no tree object associated with it.
-  Map<String, MutablePackageTree>? _packageChildren;
+  /// A list of all packages.
+  final List<SimplePackage> _packages = [];
 
   @override
   Iterable<Package> get allPackages sync* {
-    for (var package in packages) {
+    for (var package in _packages) {
       yield package;
     }
-    var children = _packageChildren;
-    if (children != null) {
-      for (var tree in children.values) {
-        yield* tree.allPackages;
+  }
+
+  bool _checkConflict(_PackageTrieNode node, SimplePackage newPackage,
+      void Function(Object error) onError) {
+    var existingPackage = node.package;
+    if (existingPackage != null) {
+      // Trying to add package that is inside the existing package.
+      // 1) If it's an exact match it's not allowed (i.e. the roots can't be
+      //    the same).
+      if (newPackage.root.path.length == existingPackage.root.path.length) {
+        onError(ConflictException(
+            newPackage, existingPackage, ConflictType.sameRoots));
+        return true;
+      }
+      // 2) The existing package has a packageUriRoot thats inside the
+      //    root of the new package.
+      if (_beginsWith(0, newPackage.root.toString(),
+          existingPackage.packageUriRoot.toString())) {
+        onError(ConflictException(
+            newPackage, existingPackage, ConflictType.interleaving));
+        return true;
+      }
+
+      // For internal reasons we allow this (for now). One should still never do
```

Comment:
We previously disallowed this.

Why are we opening this up? Is there an internal bug with details?

Do we have an issue filed to  (re)close this hole? Will it be a breaking change to do so?

Reply:
We did not disallow this. We probably thought we did, but we didn't.

This was basically a bug I tried to fix at the same time as the fix for the O(n^2), but as it turns out it's used internally (probably a bug in an auto-generated .packages file, but who knows) so I had to allow it here too. At least it's explicit.

I haven't filed an issue.